ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
+          missedCount = (ts - prev - 1) / configs.heartbeatSendIntervalMs;
+        } else {
+          missedCount = 0;
+        }
+        prev = ts;
+      }
+      // Check frame from last received heartbeat to window end
+      if (windowEnd - prev - 1 > 0) {
+        missedCount = (windowEnd - prev - 1) / configs.heartbeatSendIntervalMs;
+      }
+
+      if (missedCount < configs.heartbeatMissedThreshold) {
+        aliveHosts.put(host.toString() , true);
+
+      } else {
+        aliveHosts.put(host.toString(), false);
+      }
+    }
+  }
+
+  class SendHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      //broadcast heartbeat
+      remoteHostsURI.forEach((uri, bool) -> serviceContext.getKsqlClient().makeHeartbeatRequest(
+          uri, localHostInfo, System.currentTimeMillis()));
+    }
+  }
+
+  class DiscoverClusterTask implements Runnable {
```

Comment:
Is there a window between when a new host comes online and when this would discover them when we would't consider them alive and possibly fail the pull request?  Is there not a callback methodology (vs the polling here) so that we wouldn't have to manually configure this window?

Reply:
Yes, this can happen. I am not aware of an existing callback mechanism. There is an ongoing conversation in KLIP-12 about cluster membership. Will need to investigate this further.

Reply:
We could implement the Kafka Streams state listener down the line. https://kafka.apache.org/23/javadoc/org/apache/kafka/streams/KafkaStreams.StateListener.html to trigger the polling

Reply:
Not sure this listener works as it gets triggered on state changes of an existing streams instance. We need to listen to the creation of new instances ( persistent queries ).

Reply:
you are right. The callbacks may fire only if the changes are relevant to that instance, seems like.